addons/stock/models/stock_quant.py

```diff
@@ -378,6 +472,58 @@ def _onchange_serial_number(self):
+                    quant.company_id).property_stock_inventory, quant.location_id)
+            else:
+                move_vals = quant._get_inventory_move_values(-quant.inventory_diff_quantity, quant.location_id, quant.product_id.with_company(
+                    quant.company_id).property_stock_inventory, out=True)
+            moves.add(quant.env['stock.move'].with_context(
+                inventory_mode=False).create(move_vals).id)
+        self.env['stock.move'].browse(moves)._action_done()
+        self.location_id.write({'last_inventory_date': fields.Date.today()})
+        # next inventory date depends on if location has a cyclic inventory date
+        locations = self.mapped('location_id')
+        for location in locations:
+            next_inventory_date = False
+            if location.next_inventory_date:
+                next_inventory_date = location.next_inventory_date
+            elif location.company_id.annual_inventory_date:
+                next_inventory_date = fields.Date.today().replace(month=location.company_id.annual_inventory_date.month,
+                                                                  day=location.company_id.annual_inventory_date.day)
+                if next_inventory_date == fields.Date.today():
+                    next_inventory_date = fields.Date.today().replace(year=next_inventory_date.year + 1)
+            self.filtered(lambda q: q.location_id == location).write({'inventory_quantity': 0, 'inventory_date': next_inventory_date})
+        self.write({'inventory_diff_quantity': 0})
+
+    def _set_inventory_date(self, reset_inventory_quantity=False):
+        """ A quants next inventory date is decided on:
+        1. Does its location have a cyclic inventory set?
+        2. If not 1, then is there an annual inventory date set (by company)?
+        3. If not 1 and 2, then quant has no next inventory date.
+
+        :param reset_inventory_quantity: boolean expected to be true only when applying an inventory quantity
+        """
+        vals = {}
+        if reset_inventory_quantity:
+            vals['inventory_quantity'] = 0
+        locations = self.mapped('location_id')
+        for location in locations:
+            next_inventory_date = False
+            if location.next_inventory_date:
+                next_inventory_date = location.next_inventory_date
+            elif location.company_id.annual_inventory_date:
+                next_inventory_date = fields.Date.today().replace(month=location.company_id.annual_inventory_date.month,
+                                                                  day=location.company_id.annual_inventory_date.day)
+                if next_inventory_date == fields.Date.today():
+                    next_inventory_date = fields.Date.today().replace(year=next_inventory_date.year + 1)
+            vals['inventory_date'] = next_inventory_date
+            self.filtered(lambda q: q.location_id == location).write(vals)
```

Comment:
`write` should not be called in `_compute` method. Prefer `=` instead

Reply:
I added this method to `_apply_inventory`. I think now we have to use write? If it's better I can get re-write this method to take a location and return the `next_inventory_date` so we reuse some of the code...

Reply:
Moved the bulk of this over to stock.inventory. `_compute` now uses `=` while `_apply_inventory` still uses `write`